link-grammar/fast-match.c

```diff
@@ -203,6 +204,59 @@ static Match_node * add_to_left_table_list(Match_node * m, Match_node * l)
 }
 
 /**
+ * Validate that the uppercase part of two connectors are the same.
+ * Return true if so, false otherwise.
+ * FIXME: Use connector enumeration.
+ */
+static bool con_uc_eq(const Connector *c1, const Connector *c2)
+{
+
+	if (string_set_cmp(c1->string, c2->string)) return true;
+	if (c1->hash != c2->hash) return false;
+	if (c1->uc_length != c1->uc_length) return false;
+
+	const char *uc1 = &c1->string[c1->uc_start];
+	const char *uc2 = &c2->string[c2->uc_start];
+	/* We arrive here for less than 50% of the cases for "en" and
+	 * less then 20% of the cases for "ru", and the following
+	 * condition is always true.
+	 * How come 2 different uc strings never have the same hash value? */
```

Comment:
well, the other thing to note here is that although you are storing 16 bits of hash, there are NOT several K of UC strings -- there are only 5 or 10 of these strings -- maybe dozens at most -- since you are storing only those connectors that actually happen for some fixed word in the sentence, and typically, there simply won't be very many on any given word .. so you are hashing 5 or ten things into a space of 65536 so a hash collision would be very very unlikely.